docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+- New `QueryContainer` functionality MUST NOT be introduced but through the advanced [Entity Manager](#entity-manager) and [Repository](#repository) pattern.
+
+<details><summary markdown='span'>Additional Conventions for Project Development</summary>
+- `QueryContainers` can be used/developed further, but it is recommended to transition toward the [Entity Manager](#entity-manager) and [Repository](#repository) pattern.
+</details>
+
+## Query Object
+
+```
+src
+├── Orm
+│   └── Zed   
+│       └── [DomainName]   
+│           ├── Base
+│           │   └── [EntityName]Query.php 
+│           └── [EntityName]Query.php
+│
+└── [Organisation]
+    └── Zed
+        └── [ModuleName]
+            └── Persistence
+                └── Propel
+                    └── Abstract[EntityName]Query.php
+```
+
+**Description**
+
+Enables to write queries to the related table in an SQL engine agnostic way.
+`Query Objects` can be instantiated and used only from the [Repository](#repository) and [Entity Manager](#entity-manager) of the definer module(s).
+
+See [Propel Documentation - Query Class](#https://propelorm.org/documentation/reference/model-criteria.html)
+See more details on the 3-tier class hierarchy in [Entity](#entity), and domains in [Persistence Schema](#persistence-schema).
+
+**Convention**
+
+- Hidden, hard dependencies appearing through `join` MUST be defined via `@module [RemoteModule1][,[...]]` tag (see example below).
+
+**Example**
+
+In the below example, the `SpyProductQuery` is part of the `Product` `domain` while `SpyUrl` belongs to `Url` `domain` and `SpyProductComment` belongs to `ProductComment` `domain`.
+
+```php
+  /**
+   * @module Url,ProductComment
+   */
+  public function exampleMethod()
+  {
+      $query = new SpyProductQuery();
+      $query
+        ->joinSpyUrl()
+        ->joinSpyProductComment();
+      ...
+  }
+```
+
+
+## Repository
+
+```
+[Organization]
+└── Zed
+    └── [Module]
+        └── Persistence
+            ├── [Module]Repository.php      
+            └── [Module]RepositoryInterface.php
+```
+
+**Descriptions**
+
+Responsible for retrieving data from database by executing queries and returning the results as [Transfer Objects](#transfer-object) or native types.
+The `Repository` can be accessed from the same module's [Communication](#communication-layer-responsibilities) and [Business Layers](#business-layer-responsibilities).
+
+**Conventions**
+
+- The `Repository` class MUST define and implement an interface (`[Model]RepositoryInterface.php`) that holds the specification of each `public` method.
+- `Public` methods MUST receive and return [Transfer Objects](#transfer-object) only.
+- The methods MUST return with a collection of items or a single item (eg: using and wrapping the results of `find()` or `findOne()`).
+- `Repositories` MUST use [Query Objects](#query-object) to retrieve data from the database.
+- `Repostiories` MUST NOT alter the data in the database, only retrieve it.
+
```

Comment:
```suggestion
<details><summary markdown='span'>Additional Conventions for Project Development</summary>
- Methods can use also native php types as input arguments or result values.
</details>

```

Reply:
I made a lot of refactoring and there is always a way not to do it

Reply:
@vitaliiivanovspryker what about CTE?
`may` means that you are allowed to use queries to gain performance.

Reply:
but i agree with the proposal. there is no reason to forbid this for project development